File: RSVPTiming.swift
import Foundation

/// RSVP timing engine — Optimal Recognition Point (ORP) and per-word pacing.
///
/// Derived from thomaskolmans/rsvp-reading (MIT). See NOTICE for the full
/// attribution chain; the MIT terms require it to travel with this file.
///
/// Parity-locked to `rsvpTiming.js`. `parity-vectors.json` is the shared golden
/// fixture both sides must reproduce exactly; run the parity check before
/// changing any threshold here.
///
/// Token conventions (shared with the tokenizer module):
///   - `"\n"`       a paragraph-break token (renders as a blank beat)
///   - `"⟩"` prefix marks the first word after a break (reorientation beat)
public enum RSVPTiming {

    // MARK: - ORP (Optimal Recognition Point)

    /// Letters AND numbers are ORP-countable; punctuation is not. A pure-number
    /// token ("2026", "12,345") would otherwise pivot on its first digit
    /// instead of being centred like a word.
    ///
    /// This must match the JS twin's `\p{L}\p{N}`, so it spans every Unicode
    /// number category — not just decimal digits. `CharacterSet.decimalDigits`
    /// is Nd-only and silently excludes superscripts (Nd vs No), which made
    /// footnote-marked words like "hierarchies.¹" pivot on a different letter
    /// than iOS. `Character.isNumber` covers Nd, Nl and No.
    private static func isORPCountable(_ char: Character) -> Bool {
        char.isLetter || char.isNumber
    }

    /// Length → ORP letter position.
    ///
    ///   1-3 letters → 1st letter      12-14 → 5th
    ///   4-5         → 2nd             15-17 → 6th
    ///   6-9         → 3rd             18+   → 7th
    ///   10-11       → 4th
    public static func getORPIndex(for word: String) -> Int {
        let clean = stripBreakMarker(word)
        let count = clean.filter(isORPCountable).count

        if count <= 3 { return 0 }
        if count <= 5 { return 1 }
        if count <= 9 { return 2 }
        if count <= 11 { return 3 }
        if count <= 14 { return 4 }
        if count <= 17 { return 5 }
        return 6
    }

    /// The actual character offset of the ORP letter, skipping non-countable
    /// characters such as leading quotes.
    ///
    /// Offsets are counted in `Character`s (not unicode scalars) so the result
    /// can be used directly with `String.index(_:offsetBy:)` — counting scalars
    /// here would drift on combining marks and emoji, and would also diverge
    /// from the JS twin, which counts UTF-16 units.
    public static func getActualORPIndex(for word: String) -> Int {
        let clean = stripBreakMarker(word)
        guard !clean.isEmpty else { return 0 }

        let orpIndex = getORPIndex(for: clean)
        var count = 0

        for (i, char) in clean.enumerated() where isORPCountable(char) {
            if count == orpIndex { return i }
            count += 1
        }

        return min(orpIndex, clean.count - 1)
    }

    /// Split a word into the three display segments around its ORP letter.
    public static func splitWordForDisplay(
        _ word: String
    ) -> (before: String, orp: String, after: String) {
        let clean = stripBreakMarker(word)
        guard !clean.isEmpty else { return ("", "", "") }

        let orpIdx = getActualORPIndex(for: clean)
        let start = clean.startIndex
        let orpIndex = clean.index(start, offsetBy: orpIdx, limitedBy: clean.endIndex)
            ?? clean.index(before: clean.endIndex)
        let afterIndex = clean.index(after: orpIndex)

        return (
            before: String(clean[start..<orpIndex]),
            orp: String(clean[orpIndex]),
            after: afterIndex < clean.endIndex ? String(clean[afterIndex...]) : ""
        )
    }

    // MARK: - Pacing

    /// Trailing characters that hide the real sentence/clause punctuation:
    /// footnote markers glued to the prior word ("theology.2"), closing quotes
    /// ('."', '?"'), and closing brackets. Without stripping these, the pause
    /// silently doesn't fire and the sentence blows past with no beat.
    private static let trailingNoise =
        "[0-9\u{00B9}\u{00B2}\u{00B3}\u{2070}-\u{2079}\"'\u{2018}\u{2019}\u{201C}\u{201D})\\]}\u{00BB}\u{203A}]+$"
    private static let trailingPunct = "[.!?;:,]+$"

    /// Attenuate punctuation pauses on short words. Every word keeps SOME
    /// pause — the rhythm lives in that beat — but a 2-letter word taking the
    /// full multiplier reads as a hitch, since the eye has already absorbed it
    /// well inside the baseline interval.
    private static func shortWordScale(_ letterLength: Int) -> Double {
        if letterLength <= 1 { return 0.3 }
        if letterLength == 2 { return 0.4 }
        if letterLength <= 5 { return 0.6 }
        if letterLength == 6 { return 0.8 }
        return 1.0
    }

    /// How long to hold a word on screen, in milliseconds.
    public static func getWordDelay(
        for word: String,
        wpm: Int,
        pauseOnPunctuation: Bool = true,
        punctuationMultiplier: Double = 2.0,
        wordLengthMultiplier: Double = 0,
        lineBreakMultiplier: Double = 3.0
    ) -> Double {
        guard !word.isEmpty else { return 60_000.0 / Double(max(1, wpm)) }
        guard wpm > 0 else { return 200 }

        var baseDelay = 60_000.0 / Double(wpm)

        if word == "\n" { return baseDelay * lineBreakMultiplier }

        let isFirstAfterBreak = word.hasPrefix("⟩")
        let clean = isFirstAfterBreak ? String(word.dropFirst()) : word

        let denoised = clean.replacingOccurrences(
            of: trailingNoise, with: "", options: .regularExpression)
        let letterLength = denoised.replacingOccurrences(
            of: trailingPunct, with: "", options: .regularExpression).count
        let scale = shortWordScale(letterLength)

        // Baseline before the reorientation beat. When a word is BOTH
        // first-after-break and punctuated, take the larger of the two pauses
        // rather than compounding them — stacking 1.5x by 2x gave a ~3x hitch.
        let preBreakBaseDelay = baseDelay

        if isFirstAfterBreak {
            baseDelay *= 1.0 + 0.5 * scale
        }

        // Long words need more time. 12+ chars is roughly two standard
        // deviations above mean English word length.
        if wordLengthMultiplier > 0 && clean.count >= 12 {
            baseDelay *= 1.0 + (wordLengthMultiplier / 100.0) * Double(clean.count - 12)
        }

        if pauseOnPunctuation {
            // Em/en-dash are semantic pause markers and count as
            // sentence-end-equivalent. Hyphen-minus inside a compound like
            // "well-known" is NOT pause-worthy.
            if let last = denoised.last, ".!?;:—–".contains(last) {
                let mult = 1.0 + (punctuationMultiplier - 1.0) * scale
                return isFirstAfterBreak
                    ? max(baseDelay, preBreakBaseDelay * mult)
                    : baseDelay * mult
            }
            if denoised.hasSuffix(",") {
                let mult = 1.0 + 0.5 * scale
                return isFirstAfterBreak
                    ? max(baseDelay, preBreakBaseDelay * mult)
                    : baseDelay * mult
            }
        }

        return baseDelay
    }

    /// Whether to pause at this index, for the pause-every-N-words setting.
    public static func shouldPauseAtWord(index: Int, pauseAfterWords: Int) -> Bool {
        guard pauseAfterWords > 0, index > 0 else { return false }
        return index % pauseAfterWords == 0
    }

    // MARK: - Formatting

    /// Remaining reading time as "M:SS".
    public static func formatTimeRemaining(wordsRemaining: Int, wpm: Int) -> String {
        guard wordsRemaining > 0, wpm > 0 else { return "0:00" }

        let totalSeconds = Int(ceil(Double(wordsRemaining) / Double(wpm) * 60.0))
        return "\(totalSeconds / 60):\(String(format: "%02d", totalSeconds % 60))"
    }

    // MARK: - Token helpers

    /// Strip the first-word-after-break marker.
    public static func stripBreakMarker(_ word: String) -> String {
        word.hasPrefix("⟩") ? String(word.dropFirst()) : word
    }

    /// Whether a token renders as a visible word (not a break beat or whitespace).
    public static func isRenderable(_ token: String) -> Bool {
        let clean = stripBreakMarker(token)
        return !clean.trimmingCharacters(in: .whitespaces).isEmpty && clean != "\n"
    }

    /// A window of words centred on the current position, for multi-word
    /// display modes. Break markers are stripped from every returned word.
    public static func extractWordFrame(
        _ allWords: [String],
        centerIndex: Int,
        frameSize: Int
    ) -> (subset: [String], centerOffset: Int) {
        guard frameSize > 1, centerIndex < allWords.count else {
            let word = centerIndex >= 0 && centerIndex < allWords.count
                ? allWords[centerIndex] : ""
            return ([stripBreakMarker(word)], 0)
        }

        let radius = frameSize / 2
        let lower = max(0, centerIndex - radius)
        let upper = min(allWords.count, centerIndex + radius + 1)

        return (allWords[lower..<upper].map(stripBreakMarker), centerIndex - lower)
    }

    /// Nearest renderable index, searching forward from `startIndex` then
    /// backward. Used when a restored reading position lands on a break token.
    public static func findNearestRenderableIndex(
        in tokens: [String],
        from startIndex: Int
    ) -> Int? {
        guard !tokens.isEmpty else { return nil }
        let clamped = max(0, min(tokens.count - 1, startIndex))

        if isRenderable(tokens[clamped]) { return clamped }

        for i in (clamped + 1)..<tokens.count where isRenderable(tokens[i]) { return i }
        for i in stride(from: clamped - 1, through: 0, by: -1)
            where isRenderable(tokens[i]) { return i }

        return nil
    }
}

File: rsvpTiming.js
/**
 * RSVP timing engine — Optimal Recognition Point (ORP) and per-word pacing.
 *
 * Derived from thomaskolmans/rsvp-reading (MIT). See NOTICE for the full
 * attribution chain; the MIT terms require it to travel with this file.
 *
 * Pure functions, no dependencies. The Swift twin in `RSVPTiming.swift` is
 * parity-locked to this file; `parity-vectors.json` is the shared golden
 * fixture both sides must reproduce exactly.
 *
 * Token conventions (shared with the tokenizer module):
 *   - '\n'          a paragraph-break token (renders as a blank beat)
 *   - '⟩' prefix    marks the first word after a break (reorientation beat)
 */

// ---------------------------------------------------------------------------
// ORP (Optimal Recognition Point)
// ---------------------------------------------------------------------------

/**
 * Length → ORP letter position. Letters AND digits count; punctuation does
 * not. A pure-number token ("2026", "12,345") has no \p{L} chars, so a
 * letter-only count would pivot it on its first digit instead of centering
 * it like a word — digits are part of the recognition point.
 *
 *   1-3 letters  → 1st letter      12-14 → 5th
 *   4-5          → 2nd             15-17 → 6th
 *   6-9          → 3rd             18+   → 7th
 *   10-11        → 4th
 *
 * @param {string} word
 * @returns {number} index of the letter to highlight
 */
export function getORPIndex(word) {
  if (!word || typeof word !== 'string') return 0;

  const cleanWord = stripBreakMarker(word);
  const len = cleanWord.replace(/[^\p{L}\p{N}]/gu, '').length;

  if (len <= 3) return 0;
  if (len <= 5) return 1;
  if (len <= 9) return 2;
  if (len <= 11) return 3;
  if (len <= 14) return 4;
  if (len <= 17) return 5;
  return 6;
}

// Pre-compiled for performance. Letters AND digits are ORP-eligible so numeric
// tokens pivot by length the same way words do (see getORPIndex).
const orpCountable = /[\p{L}\p{N}]/u;

/**
 * The actual character index of the ORP letter, skipping non-countable
 * characters such as leading quotes.
 *
 * @param {string} word
 * @returns {number}
 */
export function getActualORPIndex(word) {
  if (!word || typeof word !== 'string') return 0;

  const cleanWord = stripBreakMarker(word);
  const orpIndex = getORPIndex(cleanWord);
  let letterCount = 0;

  for (let i = 0; i < cleanWord.length; i++) {
    if (orpCountable.test(cleanWord[i])) {
      if (letterCount === orpIndex) return i;
      letterCount++;
    }
  }

  return Math.min(orpIndex, cleanWord.length - 1);
}

/**
 * Split a word into the three display segments around its ORP letter.
 *
 * @param {string} word
 * @returns {{ before: string, orp: string, after: string }}
 */
export function splitWordForDisplay(word) {
  if (!word || typeof word !== 'string') {
    return { before: '', orp: '', after: '' };
  }

  const cleanWord = stripBreakMarker(word);
  const orpIndex = getActualORPIndex(cleanWord);

  return {
    before: cleanWord.slice(0, orpIndex),
    orp: cleanWord[orpIndex] || '',
    after: cleanWord.slice(orpIndex + 1),
  };
}

// ---------------------------------------------------------------------------
// Pacing
// ---------------------------------------------------------------------------

// Trailing characters that hide the real sentence/clause punctuation:
//   1. Footnote markers — many EPUBs emit footnote refs as bare digits glued
//      to the prior word ("theology.2", "Platonism,1"), or as superscripts.
//   2. Closing quotation marks — dialogue ends in '."', '?"' etc, where the
//      terminal char is the quote, not the punctuation.
//   3. Closing brackets — "(end of phrase.)", "[note.]".
// Without stripping these, the period/comma is invisible to the tests below
// and the pause silently doesn't fire — the reader feels the sentence blow
// past with no beat.
const TRAILING_NOISE = /[0-9¹²³⁴⁵⁶⁷⁸⁹⁰"'’”\)\]\}»›]+$/;
const TRAILING_PUNCT = /[.!?;:,]+$/;

/**
 * Attenuate punctuation pauses on short words.
 *
 * Every word gets SOME pause — the rhythm of "...he ran home." lives in that
 * beat — but a 2-letter word taking the full multiplier reads as a hitch,
 * because the eye has already absorbed it well inside the baseline interval.
 *
 * @param {number} letterLength
 * @returns {number} scale in [0.3, 1]
 */
function shortWordScale(letterLength) {
  if (letterLength <= 1) return 0.3;
  if (letterLength === 2) return 0.4;
  if (letterLength <= 5) return 0.6;
  if (letterLength === 6) return 0.8;
  return 1;
}

/**
 * How long to hold a word on screen, in milliseconds.
 *
 * @param {string} word - token, possibly carrying a '⟩' break marker
 * @param {number} wordsPerMinute
 * @param {boolean} [pauseOnPunctuation=true]
 * @param {number} [punctuationMultiplier=2] - multiplier at sentence ends
 * @param {number} [wordLengthMultiplier=0] - % added per char past 12
 * @param {number} [lineBreakMultiplier=3] - multiplier for a '\n' beat
 * @returns {number} delay in ms
 */
export function getWordDelay(
  word,
  wordsPerMinute,
  pauseOnPunctuation = true,
  punctuationMultiplier = 2,
  wordLengthMultiplier = 0,
  lineBreakMultiplier = 3,
) {
  if (!word || typeof word !== 'string') return 60000 / wordsPerMinute;
  if (!wordsPerMinute || wordsPerMinute <= 0) return 200; // fallback

  let baseDelay = 60000 / wordsPerMinute;

  if (word === '\n') return baseDelay * lineBreakMultiplier;

  const isFirstAfterBreak = word.startsWith('⟩');
  const cleanWord = isFirstAfterBreak ? word.substring(1) : word;

  const denoised = cleanWord.replace(TRAILING_NOISE, '');
  const letterLength = denoised.replace(TRAILING_PUNCT, '').length;
  const scale = shortWordScale(letterLength);

  // Baseline before the reorientation beat is applied. When a word is BOTH
  // first-after-break and punctuated, take the larger of the two pauses
  // rather than compounding them — stacking 1.5x by 2x gave a ~3x hitch.
  const preBreakBaseDelay = baseDelay;

  if (isFirstAfterBreak) {
    baseDelay *= 1 + 0.5 * scale;
  }

  // Long words need more time. 12+ chars is roughly two standard deviations
  // above mean English word length.
  if (wordLengthMultiplier > 0 && cleanWord.length >= 12) {
    baseDelay *= 1 + (wordLengthMultiplier / 100) * (cleanWord.length - 12);
  }

  if (pauseOnPunctuation) {
    // Em-dash and en-dash are semantic pause markers (parenthetical breaks,
    // ranges) and count as sentence-end-equivalent. Hyphen-minus inside a
    // compound like "well-known" is NOT pause-worthy.
    if (/[.!?;:—–]$/.test(denoised)) {
      const mult = 1 + (punctuationMultiplier - 1) * scale;
      return isFirstAfterBreak
        ? Math.max(baseDelay, preBreakBaseDelay * mult)
        : baseDelay * mult;
    }
    if (/,$/.test(denoised)) {
      const mult = 1 + 0.5 * scale;
      return isFirstAfterBreak
        ? Math.max(baseDelay, preBreakBaseDelay * mult)
        : baseDelay * mult;
    }
  }

  return baseDelay;
}

/**
 * Whether to pause at this index, for the pause-every-N-words setting.
 *
 * @param {number} wordIndex - 0-based
 * @param {number} pauseAfterWords - 0 disables
 * @returns {boolean}
 */
export function shouldPauseAtWord(wordIndex, pauseAfterWords) {
  if (pauseAfterWords <= 0) return false;
  if (wordIndex <= 0) return false;
  return wordIndex % pauseAfterWords === 0;
}

// ---------------------------------------------------------------------------
// Formatting
// ---------------------------------------------------------------------------

/**
 * Remaining reading time as "M:SS".
 *
 * @param {number} remainingWords
 * @param {number} wordsPerMinute
 * @returns {string}
 */
export function formatTimeRemaining(remainingWords, wordsPerMinute) {
  if (remainingWords <= 0 || !wordsPerMinute || wordsPerMinute <= 0) {
    return '0:00';
  }

  const seconds = Math.ceil((remainingWords / wordsPerMinute) * 60);
  const mins = Math.floor(seconds / 60);
  const secs = seconds % 60;

  return `${mins}:${secs.toString().padStart(2, '0')}`;
}

// ---------------------------------------------------------------------------
// Token helpers
// ---------------------------------------------------------------------------

/**
 * Strip the first-word-after-break marker.
 *
 * @param {string} word
 * @returns {string}
 */
export function stripBreakMarker(word) {
  if (typeof word !== 'string') return '';
  return word.startsWith('⟩') ? word.substring(1) : word;
}

/**
 * Whether a token renders as a visible word (not a break beat or whitespace).
 *
 * @param {string} token
 * @returns {boolean}
 */
export function isRenderable(token) {
  if (typeof token !== 'string') return false;
  const clean = stripBreakMarker(token);
  return clean.trim().length > 0 && clean !== '\n';
}

/**
 * A window of words centred on the current position, for multi-word display
 * modes. Break markers are stripped from every returned word.
 *
 * @param {string[]} allWords
 * @param {number} centerIdx
 * @param {number} frameSize - odd numbers recommended
 * @returns {{ subset: string[], centerOffset: number }}
 */
export function extractWordFrame(allWords, centerIdx, frameSize) {
  if (frameSize <= 1 || centerIdx >= allWords.length) {
    return { subset: [stripBreakMarker(allWords[centerIdx] || '')], centerOffset: 0 };
  }

  const radius = Math.floor(frameSize / 2);
  const leftBound = Math.max(0, centerIdx - radius);
  const rightBound = Math.min(allWords.length, centerIdx + radius + 1);

  return {
    subset: allWords.slice(leftBound, rightBound).map(stripBreakMarker),
    centerOffset: centerIdx - leftBound,
  };
}

/**
 * Nearest renderable index, searching forward from `startIndex` then backward.
 * Used when a restored reading position lands on a break token.
 *
 * @param {string[]} tokens
 * @param {number} startIndex
 * @returns {number} index, or -1 if nothing is renderable
 */
export function findNearestRenderableWordIndex(tokens, startIndex) {
  if (!Array.isArray(tokens) || tokens.length === 0) return -1;

  const clamped = Math.max(
    0,
    Math.min(tokens.length - 1, Number.isFinite(startIndex) ? startIndex : 0),
  );

  if (isRenderable(tokens[clamped])) return clamped;

  for (let i = clamped + 1; i < tokens.length; i += 1) {
    if (isRenderable(tokens[i])) return i;
  }
  for (let i = clamped - 1; i >= 0; i -= 1) {
    if (isRenderable(tokens[i])) return i;
  }

  return -1;
}
